src/plugins/intel_gpu/src/graph/impls/cpu/non_max_suppression.cpp

```diff
@@ -440,6 +440,159 @@ attach_non_max_suppression_impl::attach_non_max_suppression_impl() {
+        // batch_index, class_index, box_index
+        ptr[offset + 0] = static_cast<T>(valid_input[offset + 0]);
+        ptr[offset + 1] = static_cast<T>(valid_input[offset + 1]);
+        ptr[offset + 2] = static_cast<T>(valid_input[offset + 2]);
+    }
+}
+
+template <typename T>
+void store_nms_gather_output1(stream& stream, memory::ptr mem, std::vector<T> valid_input) {
+    auto valid_input_size = valid_input.size() / 3;
+
+    mem_lock<T, mem_lock_type::write> lock(mem, stream);
+    auto ptr = lock.data();
+
+    auto output_batch = static_cast<size_t>(mem->get_layout().batch());
+    for (size_t si = 0; si < std::min(valid_input_size, output_batch); ++si) {
+        auto offset = si * 3;
+        // batch_index, class_index, score
+        ptr[offset + 0] = static_cast<T>(valid_input[offset + 0]);
+        ptr[offset + 1] = static_cast<T>(valid_input[offset + 1]);
+        ptr[offset + 2] = static_cast<T>(valid_input[offset + 2]);
+    }
+}
+
+template <typename T>
+void store_nms_gather_output2(stream& stream, memory::ptr mem, std::vector<int32_t> valid_input) {
+    auto valid_input_size = valid_input.size() / 3;
+
+    mem_lock<T, mem_lock_type::write> lock(mem, stream);
+    auto ptr = lock.data();
+    ptr[0] = static_cast<T>(valid_input_size);
+}
+
+void run_nms_gather(non_max_suppression_gather_inst& instance) {
+    auto& stream = instance.get_network().get_stream();
+
+    auto valid_input0 = get_nms_gather_input<ov::element_type_traits<data_types::i32>::value_type>(stream, instance.dep_memory_ptr(0));
+    store_nms_gather_output0<ov::element_type_traits<data_types::i32>::value_type>(stream, instance.output_memory_ptr(0), valid_input0);
+
+    if (instance.outputs_memory_count() >= 2) {
+        auto data_type = instance.dep_memory_ptr(1)->get_layout().data_type;
+
+        if (data_type == cldnn::data_types::f16) {
+            auto valid_input_f16 = get_nms_gather_input<ov::element_type_traits<data_types::f16>::value_type>(stream, instance.dep_memory_ptr(1));
+            store_nms_gather_output1<ov::element_type_traits<data_types::f16>::value_type>(stream, instance.output_memory_ptr(1), valid_input_f16);
+        } else if (data_type == cldnn::data_types::f32) {
+            auto valid_input_f32 = get_nms_gather_input<ov::element_type_traits<data_types::f32>::value_type>(stream, instance.dep_memory_ptr(1));
+            store_nms_gather_output1<ov::element_type_traits<data_types::f32>::value_type>(stream, instance.output_memory_ptr(1), valid_input_f32);
+        } else {
+            throw std::runtime_error("Non max suppression gather - unsupported second output data type");
+        }
+
+        if (instance.outputs_memory_count() == 3) {
+            store_nms_gather_output2<ov::element_type_traits<data_types::i32>::value_type>(stream, instance.output_memory_ptr(2), valid_input0);
+        }
+    }
+}
+}  // namespace
+struct non_max_suppression_gather_impl : typed_primitive_impl<non_max_suppression_gather> {
```

Comment:
Random spot: do we need to copy at gather?
Can't we just reinterpret parent memory?

Reply:
I've updated to reinterpret input mem to output mem with updated layout as you mentioned.

Reply:
Actually I think gather_nms can act as reshape operation, i.e., 
1) it just can be set as "can_be_optimized', 
2) and implement on_execute() and update_output_memory() where the output memory of gather_nms is to be updated.

Reply:
Updated.